File: XUCore/additions/NSImageAdditions.m
// 
// NSImageAdditions.m
// 
// Created by Charlie Monroe
// 
// Copyright (c) 2010-13 Fuel Collective, LLC. All rights reserved.
// 


#import "NSImageAdditions.h"

#import "FCLog.h"

#if TARGET_OS_IPHONE
	#define NSZeroRect CGRectZero
	#define NSZeroSize CGSizeZero
#endif

#if !TARGET_OS_IPHONE
@interface NSImage (Private)
-(NSImage*)_imageWithSingleImageRepOfSize:(NSSize)size;
@end

@implementation NSImage (Private)

-(NSImage*)_imageWithSingleImageRepOfSize:(NSSize)size{
	if (NSEqualSizes(NSZeroSize, size)){
		return nil;
	}
	
	//size.width /= [[NSGraphicsContext currentContext] sca]
	
	NSImage *icon;
	NSSize s = [self size];
	
	if (s.width <= size.width && s.height <= size.height){
		return self;
	}
	
	CGFloat scale = [[[[NSApp windows] firstObject] screen] backingScaleFactor];
	if (scale == 0.0){
		scale = 1.0;
	}
	size.width /= scale;
	size.height /= scale;
	
	icon = [[NSImage alloc] initWithSize:size];
	[icon lockFocus];
	
	float height = (s.height>s.width)?size.height:(size.width/s.width)*s.height;
	float width = (s.width>=s.height)?size.width:(size.height/s.height)*s.width;
	
	[self drawInRect:NSMakeRect((size.width - width)/2, (size.height - height)/2, width, height) fromRect:NSMakeRect(0, 0, s.width, s.height) operation:NSCompositeCopy fraction:1.0];
	
	[icon unlockFocus];
	
	if ([[icon representations] count] > 1 || [[icon representations] count] == 0){
		FCLog(@"image scaled with more than one rep or with none: %li", [[icon representations] count]);
	}
	
	NSBitmapImageRep *imageRep = (NSBitmapImageRep *)[[icon representations] firstObject];
	[imageRep setPixelsWide:size.width];
	[imageRep setPixelsHigh:size.height];
	
	return icon;
}

@end
#endif

@implementation NSImage (FCAdditions)

#if !TARGET_OS_IPHONE
+(NSImage*)thumbnailOfFileAtURL:(NSURL*)url withSize:(NSSize)size{
	NSNumber *number = @(size.height);
	NSDictionary *dic = [NSDictionary dictionaryWithObjectsAndKeys:(id)kCFBooleanTrue, (id)kCGImageSourceCreateThumbnailFromImageIfAbsent, number, (id)kCGImageSourceThumbnailMaxPixelSize, nil];
	
	CGImageSourceRef source = CGImageSourceCreateWithURL((__bridge CFURLRef)url, NULL);
	
	if (!source){
		return nil;
	}
	
	CGImageRef image = CGImageSourceCreateThumbnailAtIndex(source, 0, (__bridge CFDictionaryRef)dic);
	
	
	if (!image){
		CFRelease(source);
		return nil;
	}
	
	NSImage *i = [[NSImage alloc] initWithCGImage:image asBitmapImageRep:YES];
	
	CFRelease(image);
	CFRelease(source);
	
	return i;
}

-(NSImage *)blackAndWhiteImage{
	NSImageRep *rep = [(NSBitmapImageRep *)[[self representations] lastObject] bitmapImageRepByConvertingToColorSpace:[NSColorSpace deviceGrayColorSpace] renderingIntent:NSColorRenderingIntentDefault];
	NSImage *image = [[NSImage alloc] initWithSize:NSMakeSize([rep size].width, [rep size].height)];
	[image addRepresentation:rep];
	return image;
}

-(void)drawAtPoint:(CGPoint)point fromRect:(CGRect)fromRect operation:(NSCompositingOperation)op fraction:(CGFloat)delta respectFlipped:(BOOL)respectFlipped{
	NSRect rect = (NSRect){ point, [self size] };
	[self drawInRect:rect fromRect:fromRect operation:op fraction:delta respectFlipped:respectFlipped hints:nil];
}
#endif
-(void)drawCenteredInRect:(NSRect)rect{
	[self drawCenteredInRect:rect fraction:1.0];
}
-(void)drawCenteredInRect:(NSRect)rect fraction:(double)fraction{
	NSImage *image = self;
	NSSize mySize = [image size];
	
	NSRect targetRect = rect;
	
	if (mySize.width/mySize.height > rect.size.width / rect.size.height){
		// Wider
		targetRect.size.width = rect.size.width;
		targetRect.size.height = mySize.height * (rect.size.width / mySize.width);
		
		targetRect.origin.y = rect.origin.y + (rect.size.height - targetRect.size.height) / 2.0;
	}else{
		// Taller
		targetRect.size.height = rect.size.height;
		targetRect.size.width = mySize.width * (rect.size.height / mySize.height);
		
		targetRect.origin.x = rect.origin.x + (rect.size.width - targetRect.size.width) / 2.0;
	}
	
#if TARGET_OS_IPHONE
	[image drawInRect:targetRect blendMode:kCGBlendModeNormal alpha:fraction];
#else
	[image drawInRect:targetRect fromRect:NSZeroRect operation:NSCompositeSourceOver fraction:fraction respectFlipped:YES hints:nil];
#endif
	
}

#if !TARGET_OS_IPHONE
-(void)_preview:(NSMutableDictionary*)sender{
	
	NSImage *i = [self _imageWithSingleImageRepOfSize:NSMakeSize([[sender objectForKey:@"width"] floatValue], [[sender objectForKey:@"height"] floatValue])];
	if (i==nil){
		return;
	}
	
	[sender setObject:i forKey:@"result"];
	
}
-(NSImage*) initWithCGImage:(CGImageRef)cgImage asBitmapImageRep:(BOOL)asBitmapImageRep {
	if (cgImage) {
		size_t width = CGImageGetWidth(cgImage);
		size_t height = CGImageGetHeight(cgImage);
		self = [self initWithSize:NSMakeSize(width, height)];
		if (asBitmapImageRep) {
			BOOL hasAlpha = CGImageGetAlphaInfo(cgImage) == kCGImageAlphaNone ? NO : YES;
			size_t bps = 8; // hardwiring to 8 bits per sample is fine for this app's purposes
			size_t spp = hasAlpha ? 4 : 3;
			NSBitmapImageRep *bitmapImageRep = [[NSBitmapImageRep alloc] initWithBitmapDataPlanes:NULL pixelsWide:width pixelsHigh:height bitsPerSample:bps samplesPerPixel:spp hasAlpha:hasAlpha isPlanar:NO colorSpaceName:NSDeviceRGBColorSpace bitmapFormat:0 bytesPerRow:0 bitsPerPixel:0];
			
			NSGraphicsContext *bitmapContext = [NSGraphicsContext graphicsContextWithBitmapImageRep:bitmapImageRep];
			[NSGraphicsContext saveGraphicsState];
			[NSGraphicsContext setCurrentContext:bitmapContext];
			CGContextDrawImage((CGContextRef)[[NSGraphicsContext currentContext] graphicsPort], CGRectMake(0.0, 0.0, width, height), cgImage);
			[NSGraphicsContext restoreGraphicsState];
			
			[self addRepresentation:bitmapImageRep];
		} else {
			[self lockFocus];
			CGContextDrawImage((CGContextRef)[[NSGraphicsContext currentContext] graphicsPort], CGRectMake(0.0, 0.0, width, height), cgImage);
			[self unlockFocus];
		}
	} else {
		self = [self init];
	}
	return self;
}

-(NSImage*)imageWithSingleImageRepOfSize:(NSSize)size{
	NSMutableDictionary *dic = [NSMutableDictionary dictionary];
	[dic setObject:@(size.width) forKey:@"width"];
	[dic setObject:@(size.height) forKey:@"height"];
	[self performSelectorOnMainThread:@selector(_preview:) withObject:dic waitUntilDone:YES];
	return [dic objectForKey:@"result"];
}
#endif

-(NSSize)proportinallyScaledSizeForMaxSize:(NSSize)size{
	NSImage *image = self;
	NSSize mySize = [image size];
	if (mySize.width < size.width && mySize.height < size.height){
		return mySize;
	}
	
	NSSize resultSize = NSZeroSize;
	
	if (mySize.width / mySize.height > size.width / size.height){
		//Wider
		resultSize.width = size.width;
		resultSize.height = mySize.height * (size.width / mySize.width);
	}else{
		//Taller
		resultSize.height = size.height;
		resultSize.width = mySize.width * (size.height / mySize.height);
	}
	return resultSize;
}

#if !TARGET_OS_IPHONE
- (NSData* )representationForFileType: (NSBitmapImageFileType) fileType {
	NSData *temp = [self TIFFRepresentation];
	NSBitmapImageRep *bitmap = [NSBitmapImageRep imageRepWithData:temp];
	NSData *imgData = [bitmap representationUsingType:fileType properties:@{ }];
	return imgData;
}

- (NSData *)JPEGRepresentation{
	return [self representationForFileType: NSJPEGFileType];
}

- (NSData *)PNGRepresentation{
	return [self representationForFileType: NSPNGFileType];
}

- (NSData *)JPEG2000Representation{
	return [self representationForFileType: NSJPEG2000FileType];  
}

- (NSData *)GIFRepresentation{
	return [self representationForFileType: NSGIFFileType];  
}

- (NSData *)BMPRepresentation{
	return [self representationForFileType: NSBMPFileType];    
}

- (NSData *)TIFFRepresentationUsingCompression:(NSTIFFCompression)compression{
	NSData *temp = [self TIFFRepresentation];
	NSBitmapImageRep *bitmap = [NSBitmapImageRep imageRepWithData:temp];
	NSData *imgData = [bitmap representationUsingType:NSTIFFFileType properties:[NSDictionary dictionaryWithObject:@(compression) forKey:NSImageCompressionMethod]];
	return imgData;
}

- (NSData *)JPEGRepresentationUsingCompressionFactor:(int)compressionFactor progressive:(BOOL)progressive{
	NSData *temp = [self TIFFRepresentation];
	NSBitmapImageRep *bitmap = [NSBitmapImageRep imageRepWithData:temp];
	NSData *imgData = [bitmap representationUsingType:NSJPEGFileType properties:[NSDictionary dictionaryWithObjects:[NSArray arrayWithObjects:[NSNumber numberWithInt:compressionFactor],[NSNumber numberWithBool:progressive],nil] forKeys:[NSArray arrayWithObjects:NSImageCompressionFactor,NSImageProgressive,nil]]];
	return imgData;
}

- (NSData *)GIFRepresentationWithDitheredTransparency:(BOOL)dither{
	NSData *temp = [self TIFFRepresentation];
	NSBitmapImageRep *bitmap = [NSBitmapImageRep imageRepWithData:temp];
	NSData *imgData = [bitmap representationUsingType:NSGIFFileType properties:[NSDictionary dictionaryWithObject:[NSNumber numberWithBool:dither] forKey:NSImageDitherTransparency]];
	return imgData;
}

- (NSData *)PNGRepresentationInterlaced:(BOOL)interlace{
	NSData *temp = [self TIFFRepresentation];
	NSBitmapImageRep *bitmap = [NSBitmapImageRep imageRepWithData:temp];
	NSData *imgData = [bitmap representationUsingType:NSPNGFileType properties:[NSDictionary dictionaryWithObject:[NSNumber numberWithBool:interlace] forKey:NSImageInterlaced]];
	return imgData;
	
}

- (void) tileInRect:(NSRect) rect {
	NSSize size = [self size];
	NSRect destRect = NSMakeRect( rect.origin.x, rect.origin.y, size.width, size.height );
	double top = rect.origin.y + rect.size.height;
	double right = rect.origin.x + rect.size.width;
	
	// Tile vertically
	while( destRect.origin.y < top ) {
		// Tile horizontally
		while( destRect.origin.x < right ) {
			NSRect sourceRect = NSMakeRect( 0, 0, size.width, size.height );
			
			// Crop as necessary
			if( ( destRect.origin.x + destRect.size.width ) > right )
				sourceRect.size.width -= ( destRect.origin.x + destRect.size.width ) - right;
			
			if( ( destRect.origin.y + destRect.size.height ) > top )
				sourceRect.size.height -= ( destRect.origin.y + destRect.size.height ) - top;
			
			// Draw and shift
			[self drawAtPoint:destRect.origin fromRect:sourceRect operation:NSCompositeSourceOver fraction:1.0];
			destRect.origin.x += destRect.size.width;
		}
		
		destRect.origin.y += destRect.size.height;
	}
	
}

#else

-(UIImage *)imageResizedToSize:(CGSize)targetSize{
	CGSize size = [self size];
	if (size.width <= targetSize.width && size.height <= targetSize.height){
		return self;
	}
	
	CGSize newSize = CGSizeZero;
	if (size.width > size.height){
		CGFloat width = targetSize.width;
		CGFloat height = size.height * (targetSize.width / size.width);
		if (height > targetSize.height){
			height = targetSize.height;
			width = size.width * (targetSize.height / size.height);
		}
		
		newSize.width = width;
		newSize.height = height;
	}else{
		CGFloat width = size.width * (targetSize.height / size.height);
		CGFloat height = targetSize.height;
		if (width > targetSize.width){
			width = targetSize.width;
			height = size.height * (targetSize.width / size.width);
		}
		
		newSize.width = width;
		newSize.height = height;
	}
	
	UIGraphicsBeginImageContextWithOptions(newSize, NO, 0.0);
	[self drawInRect:CGRectMake(0, 0, newSize.width, newSize.height)];
	UIImage *newImage = UIGraphicsGetImageFromCurrentImageContext();
	UIGraphicsEndImageContext();
	return newImage;
}
-(NSData*)PNGRepresentation{
	return UIImagePNGRepresentation(self);
}

#endif

@end

File: XUCore/additions/NSImageAdditions.swift
//
//  NSImageAdditions.swift
//  XUCore
//
//  Created by Charlie Monroe on 11/30/15.
//  Copyright © 2015 Charlie Monroe Software. All rights reserved.
//

import Foundation

public extension NSImage {
	
	/// Returns an image with just a single image representation of size.
	private func _imageWithSingleImageRepOfSize(_ size: CGSize) -> XUImage? {
		if size.isEmpty {
			return nil
		}
		
		let newSize = size
		let s = self.size
		if s.width <= newSize.width && s.height <= newSize.height {
			return self
		}
		
		let icon = NSImage(size: newSize)
		icon.lockFocus()
		
		let height = (s.height > s.width) ? newSize.height : (newSize.width / s.width) * s.height
		let width = (s.width >= s.height) ? newSize.width : (newSize.height / s.height) * s.width
		
		
		let fromRect = CGRect(x: (newSize.width - width) / 2.0, y: (newSize.height - height) / 2.0, width: width, height: height)
		self.draw(in: fromRect, from: CGRect(x: 0.0, y: 0.0, width: s.width, height: s.height), operation: .copy, fraction: 1.0)
		
		icon.unlockFocus()
		
		if icon.representations.count > 1 || icon.representations.count == 0 {
			XULog("image scaled with more than one rep or with none: \(icon.representations.count)")
		}
		
		let imageRep = icon.representations.first!
		imageRep.pixelsWide = Int(newSize.width)
		imageRep.pixelsHigh = Int(newSize.height)
		
		return icon
	}
	
	/// Returns a black & white copy of the image. May return nil, if the image
	/// contains no bitmap image representations, or if the conversion fails.
	public var blackAndWhiteImage: XUImage? {
		guard let rep = self.representations.first as? NSBitmapImageRep else {
			return nil
		}
		
		guard let bwRep = rep.converting(to: NSColorSpace.deviceGray, renderingIntent: .default) else {
			return nil
		}
		
		let image = NSImage(size: CGSize(width: rep.size.width, height: rep.size.height))
		image.addRepresentation(bwRep)
		return image
	}
	
	/// Draws the image at point from rect. If respectFlipped is true, the current
	/// context's flip is respected.
	public func draw(at point: CGPoint, fromRect: CGRect, operation op: NSCompositingOperation, fraction delta: CGFloat, respectFlipped: Bool) {
		var rect: CGRect = CGRect()
		rect.origin = point
		rect.size = self.size
		
		self.draw(in: rect, from: fromRect, operation: op, fraction: delta, respectFlipped: respectFlipped, hints: nil)
	}

	/// Inits with GCImageRef.
	public convenience init?(cgImage: CGImage, asBitmapImageRep: Bool) {
		let width = cgImage.width;
		let height = cgImage.height;
		
		self.init(size: CGSize(width: CGFloat(width), height: CGFloat(height)))
		
		if asBitmapImageRep {
			let hasAlpha = cgImage.alphaInfo == .none ? false : true
			let bps = 8; // hardwiring to 8 bits per sample is fine for general purposes
			let spp = hasAlpha ? 4 : 3;
			
			guard let bitmapImageRep = NSBitmapImageRep(bitmapDataPlanes: nil, pixelsWide: width, pixelsHigh: height, bitsPerSample: bps, samplesPerPixel: spp, hasAlpha: hasAlpha, isPlanar: false, colorSpaceName: NSDeviceRGBColorSpace, bitmapFormat: .alphaFirst, bytesPerRow: 0, bitsPerPixel: 0) else {
				return nil
			}
			
			guard let bitmapContext = NSGraphicsContext(bitmapImageRep: bitmapImageRep) else {
				return nil
			}
			
			NSGraphicsContext.saveGraphicsState()
			NSGraphicsContext.setCurrent(bitmapContext)
			
			NSGraphicsContext.current()!.cgContext.draw(cgImage, in: CGRect(x: 0.0, y: 0.0, width: CGFloat(width), height: CGFloat(height)))
			
			NSGraphicsContext.restoreGraphicsState()
			
			self.addRepresentation(bitmapImageRep)
		}else{
			self.lockFocus()
			NSGraphicsContext.current()!.cgContext.draw(cgImage, in: CGRect(x: 0.0, y: 0.0, width: CGFloat(width), height: CGFloat(height)))
			self.unlockFocus()
		}
	}
		
	/// Scales down the image and if it contains multiple image representations,
	/// removes those. May fail if the image is of zero size, has no image reps,
	/// or if some of the underlying calls fails.
	public func imageWithSingleImageRepresentation(ofSize size: CGSize) -> XUImage? {
		var result: XUImage? = nil
		XU_PERFORM_BLOCK_ON_MAIN_THREAD { () -> Void in
			result = self._imageWithSingleImageRepOfSize(size)
		}
		return result
	}
	
	/// Returns NSData with a bitmap image file type representation.
	public func representation(forFileType fileType: NSBitmapImageFileType, properties: [String : AnyObject] = [:]) -> Data? {
		guard let temp = self.tiffRepresentation else {
			return nil
		}
		
		let bitmap = NSBitmapImageRep(data: temp)
		let imgData = bitmap?.representation(using: fileType, properties: [:])
		return imgData
	}
	
	/// Returns a basic BMP image representation.
	public var bmpRepresentation: Data? {
		return self.representation(forFileType: .BMP)
	}
	
	/// Returns a basic GIF image representation.
	public var gifRepresentation: Data? {
		return self.representation(forFileType: .GIF)
	}
	
	/// Returns a GIF image representation
	public func gifRepresentation(withDitheredTransparency dither: Bool) -> Data? {
		return self.representation(forFileType: .GIF, properties: [NSImageDitherTransparency: dither as AnyObject])
	}
	
	/// Returns a basic JPEG image representation.
	public var jpegRepresentation: Data? {
		return self.representation(forFileType: .JPEG)
	}
	
	/// Returns a JPEG image representation with specified quality.
	public func jpegRepresentation(usingCompressionFactor compressionFactor: Int, progressive: Bool) -> Data? {
		let properties: [String : AnyObject] = [
			NSImageCompressionFactor: compressionFactor as AnyObject,
			NSImageProgressive: progressive as AnyObject
		]
		return self.representation(forFileType: .JPEG, properties: properties)
	}
	
	/// Returns a basic JPEG 2000 image representation.
	public var jpeg2000Representation: Data? {
		return self.representation(forFileType: .JPEG)
	}
	
	/// Returns a basic PNG image representation.
	public var pngRepresentation: Data? {
		return self.representation(forFileType: .PNG)
	}
	
	/// Returns a PNG image representation with interlace as defined.
	public func pngRepresentation(interlaced interlace: Bool) -> Data? {
		return self.representation(forFileType: .PNG, properties: [NSImageInterlaced : interlace as AnyObject])
	}
	
	/// Returns a TIFF image representation with defined compression.
	public func tiffRepresentation(usingCompression compression: NSTIFFCompression) -> Data? {
		return self.representation(forFileType: .TIFF, properties: [NSImageCompressionMethod: compression.rawValue as AnyObject])
	}
	
	/// Draws the image as tile in specified rect.
	public func tile(inRect rect: CGRect) {
		let size = self.size
		var destRect = CGRect(x: rect.minX, y: rect.minY, width: size.width, height: size.height)
		let top = rect.minY + rect.height
		let right = rect.minX + rect.width
		
		// Tile vertically
		while destRect.minY < top {
			// Tile horizontally
			while destRect.minX < right {
				var sourceRect = CGRect(x: 0.0, y: 0.0, width: size.width, height: size.height)
				
				// Crop as necessary
				if destRect.maxX > right {
					sourceRect.size.width -= destRect.maxX - right
				}
				
				if destRect.maxY > top {
					sourceRect.size.height -= destRect.maxY - top
				}
				
				// Draw and shift
				self.draw(at: destRect.origin, from: sourceRect, operation: .sourceOver, fraction: 1.0)
				destRect.origin.x += destRect.width
			}
			
			destRect.origin.y += destRect.height
		}
	}
	
}
